stl/inc/future

```diff
@@ -352,8 +370,13 @@ public:
         return _Ready_at_thread_exit;
     }
 
-    bool _Already_has_stored_result() const {
-        return _Has_stored_result;
+    bool _Already_has_stored_result() const { // Has a result or exception
+        return is_default_constructible<_Ty>::value ? _Has_stored_result
+                                                    : (_Has_stored_result || _Exception != nullptr);
+    }
+
+    bool _Constructed_result() const {
+        return is_default_constructible<_Ty>::value || _Has_stored_result;
     }

```

Comment:
I think `is_default_constructible_v` is better in MSVC STL. Not sure whether `if constexpr` is clearer here.

Reply:
Yes, we strongly prefer using variable templates (shorter, clearer, and better compiler throughput).

I agree that using `if constexpr` here is a less clear-cut case. I would recommend doing so as it can avoid compiler warnings, and it makes the logic a bit clearer (separating the compile-time from the runtime conditions).